tfhe/src/integer/gpu/ciphertext/mod.rs

```diff
@@ -5,25 +5,243 @@ use crate::core_crypto::gpu::vec::CudaVec;
+    }
+
+    /// ```rust
+    /// use tfhe::core_crypto::gpu::{CudaDevice, CudaStream};
+    /// use tfhe::integer::gpu::ciphertext::{CudaIntegerRadixCiphertext, CudaSignedRadixCiphertext};
+    /// use tfhe::integer::gpu::gen_keys_radix_gpu;
+    /// use tfhe::shortint::parameters::PARAM_MESSAGE_2_CARRY_2_KS_PBS;
+    ///
+    /// let gpu_index = 0;
+    /// let device = CudaDevice::new(gpu_index);
+    /// let mut stream = CudaStream::new_unchecked(device);
+    ///
+    /// // Generate the client key and the server key:
+    /// let num_blocks = 4;
+    /// let (cks, sks) = gen_keys_radix_gpu(PARAM_MESSAGE_2_CARRY_2_KS_PBS, num_blocks, &mut stream);
+    ///
+    /// let msg = 10i32;
+    /// let ct = cks.encrypt_signed(msg);
+    ///
+    /// // Copy to GPU
+    /// let d_ct = CudaSignedRadixCiphertext::from_signed_radix_ciphertext(&ct, &mut stream);
+    /// let d_ct_copied = d_ct.duplicate(&mut stream);
+    ///
+    /// let ct_copied = d_ct_copied.to_signed_radix_ciphertext(&mut stream);
+    /// let msg_copied = cks.decrypt_signed(&ct_copied);
+    ///
+    /// assert_eq!(msg, msg_copied);
+    /// ```
+    fn duplicate(&self, stream: &CudaStream) -> Self {
+        let ct = unsafe { self.duplicate_async(stream) };
+        stream.synchronize();
+        ct
+    }
+    /// # Safety
+    ///
+    /// - `stream` __must__ be synchronized to guarantee computation has finished, and inputs must
+    ///   not be dropped until stream is synchronised
+    unsafe fn duplicate_async(&self, stream: &CudaStream) -> Self {
+        let lwe_ciphertext_count = self.ciphertext.d_blocks.lwe_ciphertext_count();
+        let ciphertext_modulus = self.ciphertext.d_blocks.ciphertext_modulus();
+
+        let mut d_ct = CudaVec::new_async(self.ciphertext.d_blocks.0.d_vec.len(), stream);
+        d_ct.copy_from_gpu_async(&self.ciphertext.d_blocks.0.d_vec, stream);
+
+        let d_blocks =
+            CudaLweCiphertextList::from_cuda_vec(d_ct, lwe_ciphertext_count, ciphertext_modulus);
+
+        Self {
+            ciphertext: CudaRadixCiphertext {
+                d_blocks,
+                info: self.ciphertext.info.clone(),
+            },
+        }
+    }
+
+    fn is_equal(&self, other: &Self, stream: &CudaStream) -> bool {
+        let self_size = self.ciphertext.d_blocks.0.d_vec.len();
+        let other_size = other.ciphertext.d_blocks.0.d_vec.len();
+        let mut self_container: Vec<u64> = vec![0; self_size];
+        let mut other_container: Vec<u64> = vec![0; other_size];
+
+        unsafe {
+            self.ciphertext
+                .d_blocks
+                .0
+                .d_vec
+                .copy_to_cpu_async(self_container.as_mut_slice(), stream);
+            other
+                .ciphertext
+                .d_blocks
+                .0
+                .d_vec
+                .copy_to_cpu_async(other_container.as_mut_slice(), stream);
+        }
+        stream.synchronize();
+
+        self_container == other_container
```

Comment:
Can't this be done on GPU ?

Also since the code is the same between signed and unsigned this should be an impl of CudaRadixCiphertext
and in the trait CudaIntegerRadixCiphertext it should be a default function:

``` 
impl CudaRadixCiphertext  {
pub fn is_equal(&self, other) -> bool {
 //actual impl
}


trait CudaIntegerRadixCiphertext {
 // ...
fn is_equal(&self,, other) -> bool {
  self.as_ref().is_equal(other.as_ref())
}
}

Reply:
Yes this could be done on GPU, I'll open an issue for it.